### PR TITLE
Bug 1235679 - uPnP Test App 

### PR DESCRIPTION
**Pull request by punamdahiya**
Used Universal Media Server on mac to serve media content, http://www.universalmediaserver.com/

To test a) share media folder with media files in UMS under Navigational/Share settings. 
b) Open uPnP test App, which finds media server and plays media file shared

Devices with media server and test app should be on same network.

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9f0fbf765fe723189b317e8da3cc478ad48d057f

File: dev_apps/uPnP-test/js/app.js [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+      var data = response.data.Result;
+      debugLog(data);
+
+      if (data.indexOf('xmlns:dlna') == -1) {
+        data = data.replace('<DIDL-Lite ',
+          '<DIDL-Lite xmlns:dlna="urn:schemas-dlna-org:device-1-0" ');
+      }
+      var parser = new DOMParser();
+      var serializer = new XMLSerializer();
+      var xmlResponse = parser.parseFromString(response.data.Result, 'application/xml');
```
Comment (punamdahiya) [inline dev_apps/uPnP-test/js/app.js]: XML response containing one media file is getting rendered. Seeing XML not well formed error if response XML is big and contains more than one media files. Looking into it.

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b6fe65aca3df4b87c6e6efbaf97df5617b7430b1

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=105ae4783e514375c93bf77c551a6e249afb1bae

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8eb4ea8a9a094508b14e9f0a187b9a6471abafce

File: dev_apps/uPnP-test/index.html [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    <script type="text/javascript" src="js/lib/ssdp/ssdp.js"></script>
+    <script type="text/javascript" src="js/lib/plug.play/plug.upnp.js"></script>
+    <script type="text/javascript" src="js/lib/plug.play/ContentDirectory/plug.upnp_contentdirectory.js"></script>
+    <script type="text/javascript" src="js/app.js"></script>
+    <link href="style/app.css" rel="stylesheet">
+  </head>
+  <body role="application">
+    <section>
+      <header>
+        <h1 >UPnP Test</h1>
```
Comment (jimporter) [inline dev_apps/uPnP-test/index.html]: Nit: no space after `h1`

File: dev_apps/uPnP-test/js/app.js [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+  var audioPlayer;
+  var videoPlayer;
+  var imagePlayer;
+  var discoverButton;
+  var currentPlayer;
+  var playerToggler;
+  var folderList;
+
+  function debugLog(msg, level) {
+    if (!DEBUG && level == 'debug') {
```
Comment (jimporter) [inline dev_apps/uPnP-test/js/app.js]: We usually use `===` for stuff like this. Not a big deal, since this is a test app, but it might be worth fixing if our goal is to turn this into a real app.

File: dev_apps/uPnP-test/js/app.js [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    togglePlayer(true);
+  }
+
+  function playFile(evt) {
+    evt.preventDefault();
+    var fileType = evt.target.dataset.mime.split('/')[0].toLowerCase();
+    var fileFormat = evt.target.dataset.mime.split('/')[1].toLowerCase();
+
+    if (evt.target.href.substr(-3) == 'ogv') {
+      fileFormat = 'video/ogg';
```
Comment (jimporter) [inline dev_apps/uPnP-test/js/app.js]: This seems strange; isn't `fileFormat` normally the subtype, e.g. "ogg"?

File: dev_apps/uPnP-test/js/app.js [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    } else if (evt.target.href.substr(-3) == 'ogg') {
+      fileFormat = 'audio/ogg';
```
Comment (jimporter) [inline dev_apps/uPnP-test/js/app.js]: Ditto here.

File: dev_apps/uPnP-test/js/app.js [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+    folderId = folderId || '';
+
+    mediaServer.browse(folderId).then(function(response) {
+      if (!response || !response.data) {
+        return;
+      }
+
+      var data = response.data.Result.replace(/&/g, "&amp;")
+                                     .replace(/'/g, "&apos;");
```
Comment (jimporter) [inline dev_apps/uPnP-test/js/app.js]: Why is this necessary?

Comment (punamdahiya) [inline dev_apps/uPnP-test/js/app.js]: & needs escaping in title, as XML response when parsed was throwing XML not well formed error. Tested and we don't need to escape ', will update patch

File: dev_apps/uPnP-test/js/app.js (hunk 4 above)
Comment (punamdahiya) [inline dev_apps/uPnP-test/js/app.js]: Agreed, should change variable name to say MIME type as that's the attribute its used for respective media player.

File: dev_apps/uPnP-test/js/app.js (hunk 3 above)
Comment (punamdahiya) [inline dev_apps/uPnP-test/js/app.js]: Done

File: dev_apps/uPnP-test/index.html (hunk 2 above)
Comment (punamdahiya) [inline dev_apps/uPnP-test/index.html]: Done

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=43d3bbfd4e7188691ff994f65bb67fc553d6869b

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fa473251a29c19eda23eb3f89a2fb987cf4c9b5a

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=34c178b59d07287cc894edc79dcc8d232e08eecf

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=43d5d18829b09070f91418c64d2972b7f09721be
